mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/api/akka/route/handlers/JobArtifactRouteHandlerImpl.java

```diff
@@ -0,0 +1,69 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.mantisrx.master.api.akka.route.handlers;
+
+import static io.mantisrx.master.jobcluster.proto.BaseResponse.ResponseCode.CLIENT_ERROR;
+import static io.mantisrx.master.jobcluster.proto.BaseResponse.ResponseCode.SUCCESS;
+
+import io.mantisrx.master.jobcluster.proto.JobArtifactProto;
+import io.mantisrx.server.master.store.JobArtifact;
+import io.mantisrx.server.master.store.JobArtifactIDAlreadyExistsException;
+import io.mantisrx.server.master.store.MantisStorageProvider;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JobArtifactRouteHandlerImpl implements JobArtifactRouteHandler {
+    private static final Logger logger = LoggerFactory.getLogger(JobArtifactRouteHandlerImpl.class);
+
+    private final MantisStorageProvider mantisStorageProvider;
+
+    public JobArtifactRouteHandlerImpl(MantisStorageProvider mantisStorageProvider) {
+        this.mantisStorageProvider = mantisStorageProvider;
+    }
+
+    @Override
+    public CompletionStage<JobArtifactProto.ListJobArtifactsResponse> get(JobArtifactProto.ListJobArtifactsRequest request) {
+        return CompletableFuture.supplyAsync(() -> {
+            try {
+                final List<JobArtifact> jobArtifactList = mantisStorageProvider.listJobArtifacts();
+                return new JobArtifactProto.ListJobArtifactsResponse(request.requestId, SUCCESS, "", jobArtifactList);
+            } catch (IOException e) {
+                logger.warn("Error while fetching job artifacts. Traceback: {}", e.getMessage(), e);
+                return new JobArtifactProto.ListJobArtifactsResponse(request.requestId, CLIENT_ERROR, e.getMessage(), Collections.emptyList());
+            }
+        });
+    }
+
+    @Override
+    public CompletionStage<JobArtifactProto.CreateJobArtifactResponse> create(JobArtifactProto.CreateJobArtifactRequest request) {
+        final JobArtifact jobArtifact = request.getJobArtifact();
+        return CompletableFuture.supplyAsync(() -> {
+            try {
+                mantisStorageProvider.storeNewJobArtifact(jobArtifact);
+                return new JobArtifactProto.CreateJobArtifactResponse(request.requestId, SUCCESS, "", jobArtifact.getArtifactID());
+            } catch (IOException | JobArtifactIDAlreadyExistsException e) {
+                logger.warn("Error while storing new job artifact. Traceback: {}", e.getMessage(), e);
+                return new JobArtifactProto.CreateJobArtifactResponse(request.requestId, CLIENT_ERROR, e.getMessage(), jobArtifact.getArtifactID());
```

Comment:
"io exception" probably doesn't map to client error.

Reply:
good catch. this is server error.